### ISSUE TITLE
use lightning to measure flops instead of deepspeed

### ISSUE THREAD

**Issue by aaronkl** [whittle-org/whittle#276]
Flops are currently computed based on deepseed (https://github.com/whittle-org/whittle/blob/96638a1be73506cadc6f013811c797d591a3802e/whittle/metrics/flops.py#L19) which is a non-trivial dependency. We could replace this by functionality from lightning

**Comment by vladislavalerievich** [comment]
During FLOPs comparison tests on CUDA, significant discrepancies were observed between values reported by `Lightning` and `DeepSpeed` for certain model configurations. These inconsistencies likely stem from implementation differences in how each library profiles fused operations, attention, or normalization layers. While `Lightning` provides a fast FLOPs estimate, `DeepSpeed` may offer more fine-grained profiling for complex architectures. Additionally, FLOPs differences were found to be even larger when measured on CPU, highlighting the importance of consistent device usage during profiling.

![Image](https://github.com/user-attachments/assets/4bdde28a-45d4-4d50-966f-0c7da49fe125)

![Image](https://github.com/user-attachments/assets/c6baf1e0-9fc3-4792-a67f-34972abe2b76)

Here is the source:

```python
from __future__ import annotations

import os
from typing import Literal

import matplotlib.pyplot as plt
import numpy as np
import torch
from lightning.fabric.utilities.throughput import measure_flops
from litgpt.config import Config

from whittle.models.gpt import GPT

mlp_types = ["GptNeoxMLP", "LLaMAMLP", "GemmaMLP"]
norm_types = ["LayerNorm", "RMSNorm"]


def compute_flops_lightning(
    model: GPT,
    batch_size: int = 1,
    sequence_length: int = 512,
    metric: Literal["flops"] = "flops",
    to_device: str = "cpu",
    previous_device: str | None = None,
) -> float:
    """
    Estimates FLOPs for a GPT model using PyTorch Lightning on CPU for consistency.

    Args:
        model: The GPT model to profile.
        batch_size: The batch size for the input tensor. Defaults to 1.
        sequence_length: The sequence length for the input tensor. Defaults to 512.
        metric: Only "flops" is supported. Included for API compatibility.
        previous_device: The device to cast to after profiling. If None, stays on CPU.

    Returns:
        The estimated FLOPs for the model's forward pass.
    """
    if metric != "flops":
        raise ValueError(
            "PyTorch Lightning's measure_flops only supports FLOPs calculation, not MACs"
        )

    input_tensor = torch.randint(
        0, model.config.padded_vocab_size, (batch_size, sequence_length)
    )
    original_device = next(model.parameters()).device

    # Move to CPU for consistency with DeepSpeed
    model.eval()
    model.to(to_device)
    cuda_input = input_tensor.to(to_device)

    def forward_fn():
        return model(cuda_input)

    flops = measure_flops(model, forward_fn)

    # Restore original device
    if previous_device is not None:
        model.to(previous_device)
    else:
        model.to(original_device)

    return flops


def _create_gpt_model(mlp_type: str, norm_type: str) -> GPT:
    """Create a GPT model with specified MLP and norm types."""
    config = Config()
    config.padded_vocab_size = 512
    config.n_embd = 64
    config.intermediate_size = 64 * 4
    config.n_head = 8
    config.n_query_groups = 8
    config.head_size = 8
    config.n_layer = 2
    config.block_size = 512
    config.rope_n_elem = int(config.rotary_percentage * config.head_size)
    config.norm_eps = 1e-5
    config.lm_head_bias = True
    config.fix_head_size = True
    config.mlp_class_name = mlp_type
    config.norm_class_name = norm_type
    return GPT(config)


def _compute_flops_deepspeed(
    model: GPT,
    batch_size: int = 1,
    sequence_length: int = 512,
    to_device: str = "cpu",
    previous_device: str | None = None,
) -> float:
    from deepspeed.profiling.flops_profiler import get_model_profile

    input_tensor = torch.randint(
        0, model.config.padded_vocab_size, (batch_size, sequence_length)
    ).to(to_device)

    model.eval()
    model.to(to_device)
    flops, macs, _ = get_model_profile(
        model=model,
        args=(input_tensor,),
        print_profile=False,
        detailed=False,
        warm_up=1,
        module_depth=1,
        as_string=False,
    )

    if previous_device is not None:
        model.to(previous_device)

    return flops


def plot_flop_differences(
    output_path: str = "flops_comparison.png",
):
    lightning_flops = []
    deepspeed_flops = []
    labels = []
    device = "cuda"

    for mlp_type in mlp_types:
        for norm_type in norm_types:
            gpt = _create_gpt_model(mlp_type, norm_type)
            original_device = next(gpt.parameters()).device

            try:
                flops_lightning = compute_flops_lightning(
                    gpt, to_device=device, previous_device=original_device
                )
                flops_deepspeed = _compute_flops_deepspeed(
                    gpt, to_device=device, previous_device=original_device
                )

                lightning_flops.append(flops_lightning)
                deepspeed_flops.append(flops_deepspeed)
                labels.append(f"{mlp_type}\n{norm_type}")

                print(
                    f"{mlp_type}, {norm_type} → Lightning: {flops_lightning}, "
                    f"DeepSpeed: {flops_deepspeed}"
                )

            except ImportError:
                print(f"Skipping {mlp_type}, {norm_type} (DeepSpeed not installed)")
            except Exception as e:
                print(f"Error with {mlp_type}, {norm_type}: {e}")

    # Bar positions
    x = np.arange(len(labels))  # group locations
    width = 0.35  # width of the bars

    plt.figure(figsize=(14, 6))
    bars1 = plt.bar(
        x - width / 2, lightning_flops, width, label="Lightning", color="skyblue"
    )
    bars2 = plt.bar(
        x + width / 2, deepspeed_flops, width, label="DeepSpeed", color="orange"
    )

    # Labels and legend
    plt.ylabel("FLOPs")
    plt.title(f"FLOPs Comparison ({device}): Lightning vs DeepSpeed")
    plt.xticks(x, labels, rotation=45, ha="right")
    plt.legend()
    plt.tight_layout()

    # Add values on top of bars
    def annotate_bars(bars):
        for bar in bars:
            height = bar.get_height()
            plt.text(
                bar.get_x() + bar.get_width() / 2,
                height,
                f"{int(height):,}",
                ha="center",
                va="bottom",
                fontsize=8,
            )

    annotate_bars(bars1)
    annotate_bars(bars2)

    # Save and show
    plt.savefig(output_path)
    print(f"Plot saved to: {os.path.abspath(output_path)}")
    plt.show()


if __name__ == "__main__":
    plot_flop_differences()

```

**Comment by aaronkl** [comment]
I guess it's fine to stick to lightning to remove the dependency on deepspeed. Import is just that the ranking across different architectures / sub-networks is consistent, not so much the exact numbers.
